src/main/java/seedu/address/logic/commands/room/EditRoomCommand.java

```diff
@@ -0,0 +1,220 @@
+
+    private final Integer roomNumberToEdit;
+    private final EditRoomDescriptor editRoomDescriptor;
+
+    /**
+     * Constructs an EditRoomCommand to edit the room with the room number {@code Integer}.
+     *
+     * @param roomNumberToEdit room number to edit
+     * @param editRoomDescriptor details to edit the room with
+     */
+    public EditRoomCommand(Integer roomNumberToEdit, EditRoomDescriptor editRoomDescriptor) {
+        requireAllNonNull(roomNumberToEdit, editRoomDescriptor);
+        this.roomNumberToEdit = roomNumberToEdit;
+        this.editRoomDescriptor = new EditRoomDescriptor(editRoomDescriptor);
+    }
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+
+        List<Room> lastShownList = model.getModifiableRoomList().getRoomObservableList();
+        Index index = model.getModifiableRoomList().checkIfRoomPresent(roomNumberToEdit);
+
+        if (index.getZeroBased() == 0) {
+            throw new CommandException(MESSAGE_INVALID_ROOM_NOT_FOUND);
+        }
+
+        Room roomToEdit = lastShownList.get(index.getZeroBased() - 1);
+        Room editedRoom = createEditedRoom(model, roomToEdit, editRoomDescriptor);
+
+        if (!roomToEdit.isSameRoom(editedRoom) && model.containsRoom(editedRoom)) {
+            throw new CommandException(MESSAGE_DUPLICATE_ROOM);
+        }
+
+        model.setSingleRoom(roomToEdit, editedRoom);
+        return new CommandResult(String.format(MESSAGE_EDIT_ROOM_SUCCESS, editedRoom));
+    }
+
+    /**
+     * Creates and returns a {@code Room} with the details of {@code roomToEdit}
+     * edited with {@code editRoomDescriptor}.
+     *
+     * @param model Current model
+     * @param roomToEdit Room that is to be edited.
+     * @param editRoomDescriptor Details to edit the room with.
+     * @return Room that has been edited.
+     */
+    private static Room createEditedRoom(Model model, Room roomToEdit,
+                                         EditRoomDescriptor editRoomDescriptor) throws CommandException {
+        assert (roomToEdit != null);
+        int updatedRoomNumber = editRoomDescriptor.getRoomNumber().orElse(roomToEdit.getRoomNumber());
+        boolean clearRoom = (editRoomDescriptor.getIsOccupied().orElse(null) != null); //non-null just means clear room
+        boolean hasNewPatient = (editRoomDescriptor.getPatientName().orElse(null) != null);
+        boolean alreadyOccupied = roomToEdit.isOccupied();
+
+        if (clearRoom && !hasNewPatient) {
+            //case 1: clear room
+            return new Room(updatedRoomNumber);
+        } else if (!clearRoom && !hasNewPatient && alreadyOccupied) {
+            //case 2: change room number in a room with patient already
+            return new Room(updatedRoomNumber, roomToEdit.getPatient());
+        } else if (!clearRoom && !hasNewPatient) {
+            //case 3: change room number in an empty room
```

Comment:
Would it be possible to combine case 1 and case 3 since both return `Room(updatedRoomNumber)`? 🤔

`!hasNewPatient` could be the condition.

Reply:
Actually, I'm curious about why `clearRoom` is needed as a condition.

Reply:
While not entirely linked to this PR, I think we should modify `Room` to store an `Optional<Patient>`. Then, we can remove the `isOccupied` boolean and use `Optional<Patient>.isPresent()` to find out whether a room is occupied.

Reply:
yup, i just changed it!